traffic_ops/app/lib/API/Tenant.pm

```diff
@@ -0,0 +1,225 @@
+		return $self->alert("Tenant name is required.");
+	}
+	
+	if ( $params->{name} ne $self->getTenantName($id) ) {
+	        my $name = $params->{name};
+		my $existing = $self->db->resultset('Tenant')->search( { name => $name } )->get_column('name')->single();
+		if ($existing) {
+			return $self->alert("A tenant with name \"$name\" already exists.");
+		}	
+	}	
+
+	if ( !defined( $params->{parentId}) && !$self->isRootTenant($id) ) {
+		return $self->alert("Parent Id is required.");
+	}
+	
+	my $is_active = $params->{active};
+	
+	if ( !$params->{active} && $self->isRootTenant($id)) {
+		return $self->alert("Root user cannot be in-active.");
+	}
+	
+
+	if ( !defined($params->{parentId}) && !isRootTenant($id) ) {
+		return $self->alert("Only the \"root\" tenant can have no parent.");
+	}
+	
+	my $values = {
+		name      => $params->{name},
+		active    => $params->{active},
+		parent_id => $params->{parentId}
+	};
+
+	my $rs = $tenant->update($values);
+	if ($rs) {
+		my $response;
+		$response->{id}          = $rs->id;
+		$response->{name}        = $rs->name;
+		$response->{active}      = $rs->active;
+		$response->{parentId}    = $rs->parent_id;
+		$response->{lastUpdated} = $rs->last_updated;
+		&log( $self, "Updated Tenant name '" . $rs->name . "' for id: " . $rs->id, "APICHANGE" );
+		return $self->success( $response, "Tenant update was successful." );
+	}
+	else {
+		return $self->alert("Tenant update failed.");
+	}
+
+}
+
+
+sub create {
+	my $self   = shift;
+	my $params = $self->req->json;
+
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
+	}
+
+	my $name = $params->{name};
+	if ( !defined($name) ) {
+		return $self->alert("Tenant name is required.");
+	}
+
+	my $parent_id = $params->{parentId};
+	if ( !defined($parent_id) ) {
+		return $self->alert("Parent Id is required.");
+	}
+
+	my $existing = $self->db->resultset('Tenant')->search( { name => $name } )->get_column('name')->single();
+	if ($existing) {
+		return $self->alert("A tenant with name \"$name\" already exists.");
+	}
+
+	my $is_active = exists($params->{active})? $params->{active} : 0; #optional, if not set use default
+	
+	if ( !$is_active && !defined($parent_id)) {
+		return $self->alert("Root user cannot be in-active.");
+	}
+	
+	my $values = {
+		name 		=> $params->{name} ,
+		active		=> $is_active,
+		parent_id 	=> $params->{parentId}
+	};
+
+	my $insert = $self->db->resultset('Tenant')->create($values);
+	my $rs = $insert->insert();
+	if ($rs) {
+		my $response;
+		$response->{id}          	= $rs->id;
+		$response->{name}        	= $rs->name;
+		$response->{active}        	= $rs->active;
+		$response->{parentId}           = $rs->parent_id;
+		$response->{lastUpdated} 	= $rs->last_updated;
+
+		&log( $self, "Created Tenant name '" . $rs->name . "' for id: " . $rs->id, "APICHANGE" );
+
+		return $self->success( $response, "Tenant create was successful." );
+	}
+	else {
+		return $self->alert("Tenant create failed.");
+	}
+
+}
+
+
+sub delete {
```

Comment:
do you want to check here that you can't delete a tenant that has a child tenant? or check that you can't delete a tenant that is assigned to a ds or a cdn? and return a 400. rather than letting the db return a 500?